File: api/java/yr-runtime/src/test/java/com/yuanrong/executor/TestFunctionHandler.java
/*
 * Copyright (c) Huawei Technologies Co., Ltd. 2024-2024. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.yuanrong.executor;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;

import java.io.IOException;
import java.nio.ByteBuffer;
import java.util.ArrayList;

import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.JUnit4;
import org.powermock.api.support.membermodification.MemberModifier;
import org.powermock.core.classloader.annotations.PowerMockIgnore;
import org.powermock.modules.junit4.PowerMockRunner;
import org.powermock.modules.junit4.PowerMockRunnerDelegate;

import com.yuanrong.errorcode.ErrorCode;
import com.yuanrong.errorcode.ErrorInfo;
import com.yuanrong.libruntime.generated.Libruntime;
import com.yuanrong.libruntime.generated.Libruntime.FunctionMeta;
import com.yuanrong.serialization.Serializer;

/**
 * The type Http task test.
 *
 * @since 2024 /04/06
 */
@RunWith(PowerMockRunner.class)
@PowerMockIgnore("javax.management.*")
@PowerMockRunnerDelegate(JUnit4.class)
public class TestFunctionHandler {
    /**
     * Description：
     *   Confirm that the code and message from 'Create' method are correct.
     * Steps:
     *   1. Sets 'Create' method to throw an Exception
     * Expectation:
     *   The code and message return by the method are correct.
     * @throws Exception
     */
    @Test
    public void testCreateExeceptionWithCause() throws Exception {
        FunctionMeta meta = FunctionMeta.newBuilder().setClassName("MockClassName").setSignature("MockSignature")
                .setFunctionName("MockMethodName").build();
        HandlerIntf handler = new FunctionHandler();
        ReturnType returnType = handler.execute(meta, Libruntime.InvokeType.CreateInstanceStateless, null);
        assertEquals(returnType.getCode(), ErrorCode.ERR_USER_FUNCTION_EXCEPTION);
        assertTrue(returnType.getErrorInfo().getErrorMessage().contains("failed to create instance due to the cause:"));
    }

    /**
     * Description：
     *   The message should be in detailed when runtime failed to find Method.
     * Steps:
     *   1. Sets a 'MockClass' in classCache.
     *   2. Sets an invalid Signature for the 'mockMethod' in 'MockClass'.
     *   3. Calls 'invoke' method in 'FunctionHandler'.
     * Expectation:
     *   The message in returnType are in detailed.
     *
     * @throws Exception
     */
    @Test
    public void testInvokeWithFailedtoFindMethod() throws Exception {
        FunctionHandler handler = new FunctionHandler();
        handler.classCache.put("MockClass", MockClass.class);

        FunctionMeta meta = FunctionMeta.newBuilder()
                .setClassName("MockClass")
                .setFunctionName("mockMethod")
                .setSignature("invalidSignature")
                .build();
        ReturnType returnType = handler.execute(meta, Libruntime.InvokeType.InvokeFunctionStateless, null);
        assertTrue(returnType.getErrorInfo().getErrorMessage().contains("Failed to find user-definded method:"));
    }

    /**
     * Description：
     *   The message should be in detailed when runtime failed to find Method.
     * Steps:
     *   1. Sets a 'MockClass' in classCache.
     *   2. Sets the 'mockMethodwithException' in 'MockClass' throwing an exception.
     *   3. Calls 'invoke' method in 'FunctionHandler'.
     * Expectation:
     *   The Code in returnType should be ERR_USER_FUNCTION_EXCEPTION(2002).
     *
     * @throws Exception
     */
    @Test
    public void testInvokeWithUserFuncFailure() throws Exception {
        FunctionHandler handler = new FunctionHandler();
        handler.classCache.put("MockClass", MockClass.class);
        MemberModifier.field(FunctionHandler.class, "instance").set(FunctionHandler.class, new MockClass());
        FunctionMeta meta = FunctionMeta.newBuilder()
                .setClassName("MockClass")
                .setFunctionName("mockMethodWithException")
                .setSignature("()V")
                .build();
        ArrayList<ByteBuffer> args = new ArrayList<ByteBuffer>();
        ReturnType returnType = handler.execute(meta, Libruntime.InvokeType.InvokeFunction, args);
        assertEquals(ErrorCode.ERR_USER_FUNCTION_EXCEPTION, returnType.getCode());
    }

    /**
     * Description：
     *   The errorInfo of the returnType when invoke udf with no exception should have
     *   empty message.
     * Steps:
     *   1. Sets a 'MockClass' in classCache.
     *   2. Invokes the 'mockMethod' in 'MockClass'.
     *   3. Calls 'invoke' method in 'FunctionHandler'.
     * Expectation:
     *   The message in returnType should be an empty string.
     *
     * @throws Exception
     */
    @Test
    public void testReturnTypeMesgEmpty() throws Exception {
        FunctionHandler handler = new FunctionHandler();
        handler.classCache.put("MockClass", MockClass.class);
        MemberModifier.field(FunctionHandler.class, "instance").set(FunctionHandler.class, new MockClass());
        FunctionMeta meta = FunctionMeta.newBuilder()
                .setClassName("MockClass")
                .setFunctionName("mockMethod")
                .setSignature("()V")
                .build();
        ArrayList<ByteBuffer> args = new ArrayList<ByteBuffer>();
        ReturnType returnType = handler.execute(meta, Libruntime.InvokeType.InvokeFunction, args);
        assertEquals("", returnType.getErrorInfo().getErrorMessage());
    }

    @Test
    public void loadInstanceMismatchedInputException() throws ClassNotFoundException, IOException {
        String clzName = "MockClass";
        FunctionHandler.classCache.put(clzName, MockClass.class);
        byte[] clzNameBytes = Serializer.serialize(clzName);

        MockClass mockClass = new MockClass();
        byte[] clzByte = Serializer.serialize(mockClass);

        boolean isException = false;
        FunctionHandler handler = new FunctionHandler();
        try {
            handler.loadInstance(clzByte, clzNameBytes);
        } catch (IOException e) {
            isException = true;
            assertTrue(e.toString().contains("Catch MismatchedInputException while running LoadInstance"));
        }
        assertTrue(isException);
    }

    @Test
    public void testDumpInstance() {
        boolean isException = false;
        FunctionHandler handler = new FunctionHandler();
        try {
            handler.dumpInstance("testID");
        } catch (Exception e) {
            isException = true;
        }
        assertFalse(isException);
    }

    @Test
    public void testShutdown() throws Exception {
        boolean isException = false;
        FunctionHandler handler = new FunctionHandler();
        handler.classCache.put("MockClass", MockClass.class);
        MemberModifier.field(FunctionHandler.class, "instance").set(FunctionHandler.class, new MockClass());
        MemberModifier.field(FunctionHandler.class, "instanceClassName").set(FunctionHandler.class, "MockClass");
        try {
            handler.shutdown(10);
        } catch (Exception e) {
            isException = true;
        }
        assertFalse(isException);
    }

    @Test
    public void testShutdownFailed() throws Exception {
        FunctionHandler handler = new FunctionHandler();
        MemberModifier.field(FunctionHandler.class, "instance").set(FunctionHandler.class, null);
        ErrorInfo err = handler.shutdown(10);
        assertTrue(err.getErrorMessage().contains("Failed to invoke instance function"));
    }

    @Test
    public void testRecover() throws Exception {
        boolean isException = false;
        FunctionHandler handler = new FunctionHandler();
        handler.classCache.put("MockClass", MockClass.class);
        MemberModifier.field(FunctionHandler.class, "instance").set(FunctionHandler.class, new MockClass());
        MemberModifier.field(FunctionHandler.class, "instanceClassName").set(FunctionHandler.class, "MockClass");
        try {
            handler.recover();
        } catch (Exception e) {
            isException = true;
        }
        assertFalse(isException);
    }

    @Test
    public void testRecoverFailed() throws Exception {
        FunctionHandler handler = new FunctionHandler();
        MemberModifier.field(FunctionHandler.class, "instance").set(FunctionHandler.class, null);
        ErrorInfo err = handler.recover();
        assertTrue(err.getErrorMessage().contains("Failed to invoke instance function"));
    }
}

File: api/python/yr/tests/test_function_handler.py
#!/usr/bin/env python3
# coding=UTF-8
# Copyright (c) Huawei Technologies Co., Ltd. 2025. All rights reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import asyncio
import logging
import inspect
from unittest import TestCase, main
from unittest.mock import Mock, patch
from yr.executor.function_handler import FunctionHandler
from yr.code_manager import CodeManager
from yr.serialization import Serialization
from yr.err_type import ErrorCode
from yr.libruntime_pb2 import FunctionMeta, LanguageType, InvokeType, ApiType

logger = logging.getLogger(__name__)


class TestFunctionExecutor(TestCase):
    def setUp(self):
        self.handler = FunctionHandler()
        self.function_meta = FunctionMeta(applicationName="mytest",
                                          moduleName="test.py",
                                          className="counter",
                                          functionName="add",
                                          language=LanguageType.Python,
                                          codeID="123456",
                                          apiType=ApiType.Function,
                                          signature="",
                                          name="",
                                          ns="",
                                          initializerCodeID="",
                                          isGenerator=False,
                                          isAsync=False)

    @patch("yr.log.get_logger")
    @patch.object(CodeManager(), 'load_code')
    def test_execute_function(self, mock_load_code, mock_logger):
        mock_logger.return_value = logger

        class Counter:
            def __init__(self):
                self.value = 0

            def add(self, number):
                self.value = + number
                return self.value

            def get(self):
                return self.value

            def get_exception(self):
                raise RuntimeError("test user exception")

            def __yr_shutdown__(self, timeout):
                return

        mock_load_code.return_value = Counter

        Serialization._instance[Serialization._cls] = Mock()
        Serialization().deserialize.return_value = []

        # case1 CreateInstance
        _, err = self.handler.execute_function(self.function_meta, [], InvokeType.CreateInstance, 0, False)
        self.assertTrue(err.error_code == ErrorCode.ERR_OK, err.error_code)

        # case2 InvokeFunction
        # case2.1 InvokeFunction with args
        self.function_meta.functionName = 'add'
        Serialization().deserialize.return_value = [b"__YR_PLACEHOLDER__", 10]
        res, err = self.handler.execute_function(self.function_meta, [], InvokeType.InvokeFunction, 1, False)
        self.assertTrue(err.error_code == ErrorCode.ERR_OK, err.error_code)
        self.assertEqual(res[0], 10, res[0])

        # case2.2 InvokeFunction without args
        self.function_meta.functionName = 'get'
        Serialization().deserialize.return_value = []
        res, err = self.handler.execute_function(self.function_meta, [], InvokeType.InvokeFunction, 1, False)
        self.assertTrue(err.error_code == ErrorCode.ERR_OK, err.error_code)
        self.assertEqual(res[0], 10, res[0])

        # case2.3 InvokeFunction get exception
        self.function_meta.functionName = 'nothisfunction'
        Serialization().deserialize.return_value = []
        _, err = self.handler.execute_function(self.function_meta, [], InvokeType.InvokeFunction, 1, False)
        self.assertTrue(err.error_code == ErrorCode.ERR_USER_FUNCTION_EXCEPTION, err.error_code)

        # case2.4 InvokeFunction without function
        self.function_meta.functionName = 'get_exception'
        Serialization().deserialize.return_value = []
        _, err = self.handler.execute_function(self.function_meta, [], InvokeType.InvokeFunction, 1, False)
        self.assertTrue(err.error_code == ErrorCode.ERR_USER_FUNCTION_EXCEPTION, err.error_code)
        self.assertTrue("test user exception" in err.msg, err.error_code)

        # case3 CreateInstanceStateless
        res, err = self.handler.execute_function(self.function_meta, [], InvokeType.CreateInstanceStateless, 0, False)
        self.assertTrue(err.error_code == ErrorCode.ERR_OK, err.error_code)

        # case4 InvokeFunctionStateless
        def get():
            return "hello world"

        mock_load_code.return_value = get
        res, err = self.handler.execute_function(self.function_meta, [], InvokeType.InvokeFunctionStateless, 1, False)
        self.assertTrue(err.error_code == ErrorCode.ERR_OK, err.error_code)
        self.assertEqual(res[0], "hello world", res[0])

        # case5 GetNamedInstanceMeta
        _, err = self.handler.execute_function(self.function_meta, [], InvokeType.GetNamedInstanceMeta, 1, False)
        self.assertTrue(err.error_code == ErrorCode.ERR_OK, err.error_code)

        # case6 other invokeType
        _, err = self.handler.execute_function(self.function_meta, [], 111, 1, False)
        self.assertTrue(err.error_code == ErrorCode.ERR_EXTENSION_META_ERROR, err.error_code)

        err = self.handler.shutdown(10)
        self.assertTrue(err.error_code == ErrorCode.ERR_OK, err.error_code)

    @patch("yr.log.get_logger")
    @patch.object(CodeManager(), 'load_code')
    def test_execute_async_function(self, mock_load_code, mock_logger):
        mock_logger.return_value = logger

        class Counter:
            def __init__(self):
                self.value = 0

            async def async_get(self):
                await asyncio.sleep(0)
                return self.value

            def sync_get(self):
                return self.value

        mock_load_code.return_value = Counter

        Serialization._instance[Serialization._cls] = Mock()
        Serialization().deserialize.return_value = []

        # case1 CreateInstance
        _, err = self.handler.execute_function(self.function_meta, [], InvokeType.CreateInstance, 0, False)
        self.assertTrue(err.error_code == ErrorCode.ERR_OK, err.error_code)

        # case2 InvokeFunction
        # case2.1 InvokeFunction sync function
        self.function_meta.functionName = 'sync_get'
        self.function_meta.isAsync = False
        res, err = self.handler.execute_function(self.function_meta, [], InvokeType.InvokeFunction, 1, True)
        self.assertTrue(err.error_code == ErrorCode.ERR_OK, err.error_code)
        self.assertTrue(inspect.iscoroutine(res[0]), type(res[0]))

        # case2.2 InvokeFunction async function
        self.function_meta.functionName = 'async_get'
        self.function_meta.isAsync = True
        res, err = self.handler.execute_function(self.function_meta, [], InvokeType.InvokeFunction, 1, True)
        self.assertTrue(err.error_code == ErrorCode.ERR_OK, err.error_code)
        self.assertTrue(inspect.iscoroutine(res[0]), type(res[0]))

if __name__ == "__main__":
    main()
